Document cluster properties

nrwahl2 opened the pull request:
And use "PCMK__OPT" string constants for all of them (so far that's the only part that's done).

nrwahl2 commented:
I think I remember @kgaillot saying we're unable to use string macros inside `N_()` calls, so I avoided that for the new constants.

nrwahl2 commented:
Values that are documented as "nonnegative integer" in Pacemaker Explained are validated as positive integers in libcrmcommon. Need to fix that one in one direction or the other (or have two distinct types).

nrwahl2 commented on include/crm/msg_xml_compat.h:
```diff
@@ -109,6 +109,9 @@ extern "C" {
 #define TYPE(x) (((x) == NULL)? NULL : (const char *) ((x)->name))
 
+//! \deprecated Do not use
+#define XML_CONFIG_ATTR_RECHECK "cluster-recheck-interval"
```
All of the new symbols are internal and can't be included in the compat public header.

It's not unreasonable to make them public. However, there's probably not as much of a public use case for these as there is for, say, the CIB tags names, and we've been largely defaulting to private. Also the existing `PCMK__OPT_*` constants were private.

nrwahl2 replied:
**On the other hand**, we've been making `PCMK_META_*` constants public so far. There's about as much public use case for accessing properties as there is for accessing meta-attributes. Maybe we should consider private `PCMK__META` instead.

nrwahl2 replied:
Ah... we were already supposed to be doing that (see cc8ed479), but fdb2ead3 pulled us back to public lol.

kgaillot commented on doc/sphinx/Pacemaker_Explained/cluster-options.rst:
```diff
@@ -301,6 +301,17 @@ values, by running the ``man pacemaker-schedulerd`` and
+     - 0
+     - Maximum number of jobs that can be scheduled per node. If nonpositive or
+       invalid, double the number of cores is used as the maximum number of jobs
+       per node.
```
per-node PCMK__ENV_NODE_ACTION_LIMIT will override this

kgaillot reviewed (commented):
thanks

nrwahl2 commented on include/crm/common/options_internal.h:
```diff
@@ -61,7 +63,21 @@ void pcmk__validate_cluster_options(GHashTable *options,
+ *         otherwise
+ */
+static inline bool
+pcmk__valid_interval_spec(const char *value)
```
Oh, yeah. Surprised it compiles.

nrwahl2 commented:
Rebased on current main

nrwahl2 commented on doc/sphinx/Makefile.am:
```diff
@@ -120,20 +120,25 @@ INKSCAPE_CMD	= $(INKSCAPE) --export-dpi=90 -C
 		-e 's#%CRM_BLACKBOX_DIR%#@CRM_BLACKBOX_DIR@#g'			\
 		-e 's#%CRM_DAEMON_GROUP%#@CRM_DAEMON_GROUP@#g'			\
 		-e 's#%CRM_DAEMON_USER%#@CRM_DAEMON_USER@#g'			\
+		-e 's#%CRM_DAEMON_USER_RAW%#@CRM_DAEMON_USER_RAW@#g'		\
```
Ah, duh. Same with the others. `PCMK_GNUTLS_PRIORITIES` is the only one that was needed.

nrwahl2 commented on tools/crmadmin.c:
```diff
@@ -117,7 +117,7 @@ command_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError
     if (!strcmp(option_name, "--timeout") || !strcmp(option_name, "-t")) {
-        options.timeout = crm_parse_interval_spec(optarg);
+        pcmk__parse_interval_spec(optarg, &options.timeout);
         if (errno == EINVAL) {
```
Thanks

nrwahl2 commented on include/crm/common/options_internal.h:
```diff
@@ -125,6 +125,7 @@ bool pcmk__valid_sbd_timeout(const char *value);
 #define PCMK__ENV_SHUTDOWN_DELAY            "shutdown_delay"
 
 // Constants for cluster option names
+#define PCMK__OPT_BATCH_LIMIT               "batch-limit"
```
thanks

nrwahl2 commented on daemons/controld/controld_throttle.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2013-2021 the Pacemaker project contributors
+ * Copyright 2013-2023 the Pacemaker project contributors
```
thanks

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -480,13 +480,19 @@ crm_str_to_boolean(const char *s, int *ret)
+
+        if (ret != NULL) {
+            *ret = TRUE;
+        }
```
Done

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -412,6 +412,57 @@ crm_get_msec(const char *input)
+ *         \p input is valid or \c NULL, and \c EINVAL otherwise)
+ */
+int
+pcmk__parse_interval_spec(const char *input, guint *result_ms)
```
I can do that. I didn't see much need since for anything an external program is likely to need to parse, Pacemaker has already converted it to milliseconds or seconds (depending on the context)... unless `pcmk__parse_interval_spec()` can handle the output of `pcmk__readable_interval()`. That just occurred to me.

I felt like exposing more of Pacemaker's generic time representation stuff (much like its generic XML or string stuff) was a bad idea. I don't feel that strongly about it though.

kgaillot replied:
I agree we shouldn't expose generic time handling, but interval specs are Pacemaker-specific CIB syntax, so external code might reasonably want to parse it. The readable interval format isn't the same, but that might have been a good idea.

nrwahl2 commented:
Updates:
* Cluster option symbols are now public, besides "remove-after-stop", which is deprecated.
  * This includes existing internal cluster option symbols
* cts-cli daemons test outputs have been updated as appropriate.
* Years are now 2024.
* Added a commit on top to make `pcmk_parse_interval_spec()` public. That way it's easy to revert if you change your mind.
* Reverted unnecessary sed replacements in doc/sphinx/Makefile.am
* Noted in doc that local node-action-limit overrides cluster-wide one
* De-inlined option validators
* Added unit tests for NULL result argument in `crm_str_to_boolean()`

kgaillot commented on include/crm/msg_xml.h:
```diff
@@ -71,6 +71,59 @@ extern "C" {
 
+/*
+ * Cluster options
+ */
```
Let's put these in a new crm/common/options.h and include that from crm/crm.h

kgaillot reviewed (commented):
looks good
